File: iso27001-fastapi/app/domain/users/service.py
import hashlib
from app.domain.users.repository import UserRepositoryInterface
from app.domain.users.models import User
from app.domain.users.schemas import CreateUserRequest, UpdateUserRequest
from app.domain.users.events import UserCreated
from app.config.security import hash_password
from app.domain.exceptions import ConflictError
from app.domain.events import EventBus, event_bus

class UserService:
    def __init__(self, repo: UserRepositoryInterface, bus: EventBus = event_bus):
        self.repo = repo
        self.bus = bus

    def create_user(self, request: CreateUserRequest) -> User:
        if self.repo.exists_by_email(request.email):
            raise ConflictError("Email already exists")
        
        hashed_pw = hash_password(request.password)
        user = User(
            email=request.email,
            hashed_password=hashed_pw,
            full_name=request.full_name,
            role="viewer"
        )
        saved_user = self.repo.save(user)
        
        # A.12: Emit event for audit trail (never log raw email)
        email_hash = hashlib.sha256(str(saved_user.email).encode()).hexdigest()
        event = UserCreated(user_id=str(saved_user.id), email_hash=email_hash, role=str(saved_user.role))
        self.bus.publish(event)
        
        return saved_user

    def list_users(self, skip: int, limit: int) -> list[User]:
        return self.repo.get_all(skip, limit)

    def update_user(self, user: User, request: UpdateUserRequest) -> User:
        if request.email and request.email != user.email:
            if self.repo.exists_by_email(request.email):
                raise ConflictError("Email already exists")
            user.email = request.email  # type: ignore[assignment]

        if request.full_name:
            user.full_name = request.full_name  # type: ignore[assignment]
            
        return self.repo.save(user)

    def delete_user(self, user: User) -> None:
        self.repo.delete(user)

File: iso27001-gin/internal/domain/users/service.go
package users

import (
	"crypto/sha256"
	"fmt"

	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

// EventPublisher publishes domain events to registered listeners.
type EventPublisher interface {
	Publish(event any)
}

// Service encapsulates User domain logic.
// No framework imports — pure business logic only (DDD domain layer contract).
type Service struct {
	repo      Repository
	publisher EventPublisher
}

// NewService constructs a UserService with its required dependencies.
func NewService(repo Repository, publisher EventPublisher) *Service {
	return &Service{repo: repo, publisher: publisher}
}

// CreateUser hashes the password (bcrypt cost 12 — A.10) and persists the user.
// Emits UserCreatedEvent for the audit listener (A.12).
func (s *Service) CreateUser(email, password, role string) (*User, error) {
	exists, err := s.repo.ExistsByEmail(email)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, ErrConflict{Message: "email already exists"}
	}

	hashed, err := bcrypt.GenerateFromPassword([]byte(password), 12) // A.10: cost ≥ 12
	if err != nil {
		return nil, err
	}

	u := &User{
		Email:          email,
		HashedPassword: string(hashed),
		Role:           role,
		IsActive:       true,
	}
	saved, err := s.repo.Save(u)
	if err != nil {
		return nil, err
	}

	// A.12: emit domain event with email_hash only — never raw PII
	emailHash := fmt.Sprintf("%x", sha256.Sum256([]byte(email)))
	s.publisher.Publish(UserCreatedEvent{
		UserID:    saved.ID.String(),
		EmailHash: emailHash,
		Role:      saved.Role,
	})

	return saved, nil
}

// ListUsers returns a paginated slice of active users.
func (s *Service) ListUsers(skip, limit int) ([]*User, error) {
	return s.repo.FindAll(skip, limit)
}

// UpdateUser applies partial updates; rejects duplicate email (A.9).
func (s *Service) UpdateUser(u *User, email, fullName string) (*User, error) {
	if email != "" && email != u.Email {
		exists, err := s.repo.ExistsByEmail(email)
		if err != nil {
			return nil, err
		}
		if exists {
			return nil, ErrConflict{Message: "email already exists"}
		}
		u.Email = email
	}
	if fullName != "" {
		u.FullName = fullName
	}
	return s.repo.Save(u)
}

// DeleteUser soft-deletes a user, preserving the audit trail (A.12).
func (s *Service) DeleteUser(id string) error {
	uid, err := uuid.Parse(id)
	if err != nil {
		return err
	}
	return s.repo.SoftDelete(uid)
}

// VerifyPassword performs constant-time bcrypt comparison (A.10).
func (s *Service) VerifyPassword(plain, hashed string) bool {
	return bcrypt.CompareHashAndPassword([]byte(hashed), []byte(plain)) == nil
}

// ErrConflict signals a domain uniqueness violation.
type ErrConflict struct{ Message string }

func (e ErrConflict) Error() string { return e.Message }
